### PR TITLE
Make DOS integrations more robust to noise / band edge effects

### PR DESCRIPTION
**Pull request by kavanase**
This PR adjusts carrier concentration integrals to start from near the mid-gap position (intermediate between `idx_vbm` and `idx_cbm`), making the results now mostly independent of auto-determined VBM/CBM positions and indices (which can be sensitive to noise effects, and artificially affect trapezoidal integration by terminating integration early). 
Now it integrates the DOS with the Fermi-Dirac occupations, with the DOS being zero in the (true) gap until it reaches the (true) CBM/VBM, to give greater quantitative accuracy in the results.

This is essentially the same changes that were made to the ``pymatgen`` ``FermiDos`` integration scheme here:
https://github.com/materialsproject/pymatgen/pull/3879
https://github.com/materialsproject/pymatgen/pull/4240
– though in that case it still used a simple left Riemann / rectangular sum integration, now updated to use the more accurate trapezoidal integration as in `py-sc-fermi`:
https://github.com/materialsproject/pymatgen-core/pull/64

(And so now also has the desirable result of `doped`, `pymatgen` and `py-sc-fermi` doping calculations being fully self-consistent).

### CODE REVIEW

File: py_sc_fermi/dos.py [hunk 1]
```diff
@@ -230,15 +228,27 @@ def _p0_index(self) -> int:
         Returns:
             int: index of vbm
         """
-        return np.where(self._edos <= 0)[0][-1]
+        return int(np.argmin(np.abs(self._edos)))
```
Comment (bjmorgan) [inline py_sc_fermi/dos.py]: This now gives different behaviour if `self._edos` doesn't bracket zero. Previously, we would get `IndexError`. Now we get a silently wrong index; e.g. for an edos starting above zero, `argmin(|edos|)` returns `0`, the constructor succeeds, and downstream calculations return plausible-looking but wrong numbers.

If we do want to require that the edos brackets zero, we can validate in `__init__`:

```python
if not (self._edos[0] <= 0 <= self._edos[-1]):
    raise ValueError(…)
```

Comment (bjmorgan) [reply py_sc_fermi/dos.py]: More generally, what is the rationale for changing the band edge indexing scheme? Previously, the band edge indexes were guaranteed to be inside the VB / CB. Now we get the *closest* index, which might be inside or outside, depending on the grid values.

File: py_sc_fermi/dos.py [hunk 2]
```diff
@@ -230,15 +228,27 @@ def _p0_index(self) -> int:
...
         """
-        return np.where(self._edos >= self.bandgap)[0][0]
+        return int(np.argmin(np.abs(self._edos - self.bandgap)))
+
+    def _mid_gap_index(self) -> int:
+        """find index of the mid-gap position in ``self.edos``"""
+        return (self._p0_index() + self._n0_index()) // 2
+
+    def _p0_integration_index(self) -> int:
+        """find index of the starting point for integration of hole densities"""
```
Comment (bjmorgan) [inline py_sc_fermi/dos.py]: The docstring says this is the starting point for integrating the hole densities, but it is used as a slice end-point `[:idx]`.

File: tests/test_dos.py [hunk 3]
```diff
@@ -139,35 +140,39 @@ def test_from_vasprun_type_error_spin_separated(self):
...
+                "dos": np.array([dos_data / 2, dos_data / 2]),
+                "edos": edos,
+                "bandgap": bandgap,
                 "nelect": 10,
-                "spin_pol": True
+                "spin_pol": True,
             }
         )
         self.assertEqual(dos.nelect, 10)
         self.assertEqual(dos.bandgap, 3.0)
```
Comment (bjmorgan) [inline tests/test_dos.py]:
```suggestion
        self.assertEqual(dos.bandgap, bandgap)
```

File: tests/test_dos.py
Comment (bjmorgan) [inline tests/test_dos.py:1]: Worth adding tests that directly verify the PR's stated motivation. As it stands, the updated tests check the same properties the old tests did, just on a slightly different DOS fixture. None of them would fail on `main`, so the improvement isn't actually locked down anywhere.

The central claim that "results are now mostly independent of auto-determined VBM/CBM positions and indices" could be captured by a test along these lines:

```python
def test_carrier_concentrations_robust_to_grid_alignment(self):
    bandgap = 3.0
    nelect = 10
    e_fermi, T = 1.5, 298

    edos_a = np.linspace(-10.0, 10.0, 100)
    edos_b = edos_a + 0.05  # shifted by a fraction of a grid spacing

    dos_a = np.where((edos_a <= 0) | (edos_a >= bandgap), 1.0, 0.0)
    dos_b = np.where((edos_b <= 0) | (edos_b >= bandgap), 1.0, 0.0)

    result_a = DOS(dos=dos_a, edos=edos_a, bandgap=bandgap, nelect=nelect)\
                   .carrier_concentrations(e_fermi, T)
    result_b = DOS(dos=dos_b, edos=edos_b, bandgap=bandgap, nelect=nelect)\
                   .carrier_concentrations(e_fermi, T)

    np.testing.assert_allclose(result_a, result_b, rtol=1e-3)
```

On `main` the grid shift moves the integration termination by one grid point through a region where the DOS is finite, changing the integral by a few percent (outside `rtol=1e-3`). With this PR the integration terminates at mid-gap where the DOS is zero, so the two results should agree.

Also worth adding direct unit tests for the new `_mid_gap_index`, `_p0_integration_index`, and `_n0_integration_index` helpers, particularly an edge case with a narrow bandgap that exercises the `max(...)` / `min(...)` clamping.

File: py_sc_fermi/dos.py
Comment (bjmorgan) [inline py_sc_fermi/dos.py:1]: The integration indices (`_p0_index`, `_n0_index`, `_mid_gap_index`, `_p0_integration_index`, `_n0_integration_index`) depend only on `self._edos` and `self._bandgap`, both of which are set in `__init__` and never mutated afterwards. So they're effectively constant for the lifetime of a DOS instance, and we're recomputing them on every call, including inside the Fermi-energy solver loop (via `_p_func` / `_n_func`).

It is worth caching these in __init__ and dropping the helper methods entirely. 

Sketch:
```python
def __init__(self, dos, edos, bandgap, nelect, spin_polarised=False):
    self._edos = edos
    self._bandgap = bandgap
    ...

    # Cache integration indices — depend only on edos and bandgap,
    # neither of which can change after construction. Must be set
    # before normalise_dos(), which reads _p0_int_idx via sum_dos().
    self._p0_idx = int(np.argmin(np.abs(self._edos)))
    self._n0_idx = int(np.argmin(np.abs(self._edos - self._bandgap)))
    mid_gap = (self._p0_idx + self._n0_idx) // 2
    self._p0_int_idx = max(mid_gap, self._p0_idx + 1)
    self._n0_int_idx = min(mid_gap, self._n0_idx - 1) + 1

    self.normalise_dos()
    ...
```
Then `sum_dos, carrier_concentrations`, `_p_func`, and `_n_func` read the cached attrs directly, and the five `_*_index` methods disappear.

The existing `test__p0_index` / `test__n0_index` tests would then need to be rewritten to assert against `_p0_idx` / `_n0_idx` directly.

Alternatively, we could do this:

```python
def __init__(self, ...):
    ...
    self._p0_idx = self._compute_p0_index()
    self._n0_idx = self._compute_n0_index()
    mid_gap = (self._p0_idx + self._n0_idx) // 2
    self._p0_int_idx = max(mid_gap, self._p0_idx + 1)
    self._n0_int_idx = min(mid_gap, self._n0_idx - 1) + 1
    ...

def _compute_p0_index(self) -> int:
    return int(np.argmin(np.abs(self._edos)))

def _compute_n0_index(self) -> int:
    return int(np.argmin(np.abs(self._edos - self._bandgap)))
```
Then the band-edge definitions are clearly defined and separately testable. But these are numpy one-liners, so they can also just be inlined.

**Review by bjmorgan — COMMENTED** [review]
@kavanase thanks for this. I've made some notes and suggestions.

File: py_sc_fermi/dos.py (hunk 2 above)
Comment (kavanase) [reply py_sc_fermi/dos.py]: (Outdated)

File: py_sc_fermi/dos.py
Comment (kavanase) [reply py_sc_fermi/dos.py:1]: Resolved by [b5410f9](https://github.com/bjmorgan/py-sc-fermi/pull/55/commits/b5410f91ad278f16f2888e194d81e80611d2160f)

File: py_sc_fermi/dos.py (hunk 1 above)
Comment (kavanase) [reply py_sc_fermi/dos.py]: Bracketing catch and test added.

Comment (kavanase) [reply py_sc_fermi/dos.py]: The motivation is to make the integration result more robust / less sensitive to noise. Using near-mid-gap termination for the carrier density integrations makes the exact band-edge index immaterial (integrand is zero for zero-DOS regions throughout the gap), and more accurate (ensuring no missed band-edge density due to a band index _inside_ the VB as you say).

> Previously, the band edge indexes were guaranteed to be inside the VB / CB. Now we get the closest index, which might be inside or outside, depending on the grid values.

Yes exactly. If the DOS energies are e.g. [-0.9999, 0.0001, 1.0001], the old scheme would use -0.9999 as the VBM, while the new scheme uses 0.0001. Small noise artefacts like this (on the order of 1e-5) are possible and were seen e.g. in converting `pymatgen` DOSs /vasprun-parsed DOS to `py-sc-fermi` DOSs.

File: tests/test_dos.py
Comment (kavanase) [reply tests/test_dos.py:1]: Done! [656c358](https://github.com/bjmorgan/py-sc-fermi/pull/55/commits/656c35856aa9b54d1a40b972d05efdfc65c9b52d)

**Comment by kavanase** [comment]
I think these are all now resolved @bjmorgan

File: py_sc_fermi/dos.py (hunk 1 above)
Comment (bjmorgan) [reply py_sc_fermi/dos.py]: I'm still not clear how changing the band-edge indexing calculation interacts with the choice to use near-mid-gap termination. Aren't these orthogonal? Does the new band-edge indexing *also* reduce sensitivity to the grid offset?

Comment (bjmorgan) [reply py_sc_fermi/dos.py]: Ah wait. You mean if the energy grid accumulates noise, it becomes offset from zero, which under the old scheme could cause the band edge index to jump nearly a full grid spacing.

Comment (bjmorgan) [reply py_sc_fermi/dos.py]: Probably worth a comment describing this exact case to make it clear to future readers of the code.

Comment (kavanase) [reply py_sc_fermi/dos.py]: Done!
